PHP Elasticsearch Laravel post for Feb 9 (#1072)
oktadeveloper/okta-blog: _source/_posts/2022-02-09-a-developers-guide-to-elasticsearch-with-laravel.md

image: blog/elasticsearch-laravel/elasticsearch-laravel.jpg
type: conversion
github: https://github.com/oktadev/okta-php-laravel-elasticsearch-example
---

Elasticsearch is a search engine that can be integrated into any of your tech stacks using its REST APIs and its wide selection of libraries.

By installing and integrating Elasticsearch into your system, you can add a search engine to your website, add autocomplete functionalities, use its machine learning functionalities to perform analysis on your data and provide better results for your users, and much more.

In this tutorial, you'll learn how to integrate Elasticsearch into a Laravel blog. You'll also add authentication to your blog using Okta and allow authenticated users to create new blog posts.

## What is Elasticsearch?

[Elasticsearch](https://www.elastic.co) allows you to index your data and perform searches quickly and efficiently. It also allows you to better understand your data, and to track your query load times using tools like [Kibana](https://www.elastic.co/kibana/), a visualization dashboard for your data .

> If you have an e-commerce website, it's important to add a search engine that will allow your users to search through thousands of products, including multiple fields like titles, descriptions, and categories in the search. Elasticsearch a versatile search engine that not only makes the process easy for you, but also indexes and [caches](https://www.elastic.co/blog/elasticsearch-caching-deep-dive-boosting-query-speed-one-cache-at-a-time) data to provide a faster and more optimized response to your queries.

Reviewer: ```suggestion
If you have an e-commerce website, it's important to add a search engine that will allow your users to search through thousands of products, including multiple fields like titles, descriptions, and categories in the search. Elasticsearch is a versatile search engine that not only makes the process easy for you, but also indexes and [caches](https://www.elastic.co/blog/elasticsearch-caching-deep-dive-boosting-query-speed-one-cache-at-a-time) data to provide a faster and more optimized response to your queries.
```
typo/missing verb

{% img blog/elasticsearch-laravel/configure-app.png alt:"Configure app settings" width:"800" %}{: .center-image }

At the bottom of the form, under **Assignments**, choose **Allow everyone in your organization to access**. This allows any user in your Okta organization to log in to the blog.

Once you're done, click on **Save**. You'll be redirected to the page of the app you just created. On that page, you can see **Client ID** and **Client Secret** keys. Copy them and add them as new keys in `.env`:

```
OKTA_BASE_URL=
OKTA_CLIENT_ID=
OKTA_CLIENT_SECRET=
OKTA_REDIRECT_URI=http://localhost:8000/login/redirect
```

The last thing you need is your organization's Okta base URL, which can be found by going to **Security**, then **API**. It'll be in the table under **Issuer URI**. You only need the authority, not the full path.

{% img blog/elasticsearch-laravel/get-base-uri.png alt:"Get the base URI from the issuer URI" width:"800" %}{: .center-image }

Okta is now ready to be used for authentication!

#### Add authentication in Laravel

> Back in Laravel, you'll add the authentication with Okta by using [Socialite](https://laravel.com/docs/8.x/socialite). Socialite allows you to add social logins, like Okta, to your Laravel app.

Reviewer: ```suggestion
Back in Laravel, you'll add the authentication with Okta by using [Socialite](https://laravel.com/docs/8.x/socialite). Socialite allows you to add social logins, as well as OAuth providers like Okta, to your Laravel app.
```
more accurate description of Okta

Install the package for authentication with Okta using Socialite:

```bash
composer require laravel/socialite socialiteproviders/okta
```

After installing the package, you need to add Okta in the `config/services.php` so that Socialite can find and use its keys:

```php
'okta' => [    
        'base_url' => env('OKTA_BASE_URL'),
        'client_id' => env('OKTA_CLIENT_ID'),  
        'client_secret' => env('OKTA_CLIENT_SECRET'),  
        'redirect' => env('OKTA_REDIRECT_URI'),
        'auth_server_id' => env('OKTA_SERVER_ID')
      ],
```

You also need to configure Socialite's listener to listen to the `SocialiteWasCalled` event to use Okta. In `app/Providers/EventServiceProvider.php` add the event in the `$listen` array inside the `EventServiceProvider` class:

```php
$listen = [
  ...,
  \SocialiteProviders\Manager\SocialiteWasCalled::class => [
              \SocialiteProviders\Okta\OktaExtendSocialite::class.'@handle',
        ],
]
```

You can now use Socialite to authenticate with Okta.

> Start first by importing the `Socialite` facade at the beginning of `app/Http/Controllers/BlogController.php`:

Reviewer: ```suggestion
Start by first importing the `Socialite` facade at the beginning of `app/Http/Controllers/BlogController.php`:
```
more accurate

{% img blog/elasticsearch-laravel/okta-login-prompt.png alt:"Okta login prompt" width:"400" %}{: .center-image }

Before you can test the authentication, you need to add the redirect route that will handle the response of the successful authentication.

In `app/Http/Controllers/BlogController.php`, add the necessary import at the beginning of the file for `Auth`, which you'll be using for authentication:

```php
use Illuminate\Support\Facades\Auth;
use App\Models\User;
```

Then, add the following method inside the `BlogController` class:

```php
public function handleLogin () {
    $oktaUser = Socialite::driver('okta')->user();

    $user = User::where('email', $oktaUser->email)->first();

    if (!$user) {
        $user = User::create([
            'email' => $oktaUser->email,
            'name'  => $oktaUser->name,
            'token' => $oktaUser->token,
        ]);
    } else {
        $user->token = $oktaUser->token;
        $user->save();
    }

    try {
        Auth::login($user);
    } catch (\Throwable $e) {
        return redirect()->route('login');
    }

    return redirect('/');
}
```

> This tells Socialite to retrieve the Okta user, then checks if the user's email address is in the database. If it is, the `token` field is changed to reflect the user's token. If it isn't, a new user is created, and the `token` field updates with their information. The user is then authenticated on your Laravel blog using `Auth::login` and is redirected to the homepage.

Reviewer: ```suggestion
This tells Socialite to retrieve the Okta user, then checks if the user's email address is in the database. If it is, the `token` field is changed to reflect the user's token. If it isn't, a new user is created, and the `token` field updates with their information. The user is then authenticated on your Laravel blog using `Auth::login` and is redirected to the home page.
```
consistent two-word home page

{% img blog/elasticsearch-laravel/new-post-page.png alt:"New post page" width:"800" %}{: .center-image }

To add the form POST request handler route, add the following to `app/Http/Controllers/BlogController.php`:

{% raw %}
```php
public function savePost (Request $request) {
    //save post
    $user = Auth::user();

    $post = new Post([
        'title' => $request->get('title'),
        'content' => $request->get('content'),
        'user_id' => $user->id
    ]);
    $post->save();

    return redirect('/');
}
```
{% endraw %}

This will save the new post and redirect you back to the home page.

Add the new route in `routes/web.php`:

```php
Route::post('/post/new', [App\Http\Controllers\BlogController::class, 'savePost'])->name('post.new')->middleware('auth');
```

Change the value for the `action` attribute of the form in `post-form.blade.php`:

{% raw %}
```php
<form method="POST"  action="{{ route('post.new') }}">
```
{% endraw %}

Open the page with the **Add Post** form again, enter a title and body content, then submit the form. You'll be redirected to the home page, which will now display the post you added.

> ### Add Search functionality

Reviewer: ```suggestion
### Add search functionality
```
style guide - sentence case headers
